### PR TITLE
Torch conversion for VAD-Diarization pipeline

### PR DESCRIPTION
**Pull request by tango4j**
Signed-off-by: Taejin Park <tango4j@gmail.com>

# What does this PR do ?

This PR replaces the numpy based functions with torch based functions in VAD-Diarization pipeline. This conversion is motivated for easier conversion from NeMo system to Riva production system. Most of the functions in vad_utils.py and nmesc_clustering.py are compatible with torch.jit.script decorator.   

**Collection**: ASR

# Changelog 
- Converted all the numpy based functions in nmesc_clustering.py, vad_utils.py
- Converted numpy based signal chain in speaker_utils.py
- Changed numpy or python list initialization to torch in clustering_diarizer.py

# Usage
Please refer to [speaker diarization part of the NeMo docs](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/speaker_diarization/datasets.html).

```python
python examples/speaker_tasks/diarization/offline_diarization.py \
    diarizer.manifest_filepath="/path/to/ch109_input_manifest.json" \
    diarizer.oracle_num_speakers=2  \ # Since there are exactly 2 speakers per each CH109 session
    diarizer.oracle_vad=True \ # Use oracle VAD extracted from RTTM files.
    diarizer.collar=0.25 \
    diarizer.ignore_overlap=True \
    diarizer.speaker_embeddings.model_path="titanet_large" \
    diarizer.speaker_embeddings.window_length_in_sec=[1.5,1.0,0.5] # Multiscale setting \
    diarizer.speaker_embeddings.shift_length_in_sec=[0.75,0.5,0.25] # Multiscale setting \
    diarizer.speaker_embeddings.parameters.multiscale_weights=[0.33,0.33,0.33] # Equal weights 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [  ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

ASR collection contributors.

### CODE REVIEW

File: nemo/collections/asr/parts/utils/diarization_utils.py [hunk 1]
```diff
@@ -538,7 +538,11 @@ def make_json_output(self, uniq_id, diar_hyp, word_dict_seq_list, total_riva_dic
             total_riva_dict[uniq_id] = riva_dict
             prev_speaker = speaker
 
+        # note that we need to add the very last sentence.
```
Comment (tango4j) [inline nemo/collections/asr/parts/utils/diarization_utils.py:541]: Just to inform other reviewers, this change is made because in the recent PR(https://github.com/NVIDIA/NeMo/pull/3897), [demsarjure](https://github.com/demsarjure) missed to put the very last line for transcription generation so I add it here.

File: nemo/collections/asr/parts/utils/vad_utils.py [hunk 2]
```diff
@@ -92,15 +94,16 @@ def prepare_manifest(config):
     return manifest_vad_input
 
 
-def write_vad_infer_manifest(file, args_func):
+def write_vad_infer_manifest(file: dict, args_func: dict) -> list:
     """
     Used by prepare_manifest.
     Given a list of files, write them to manifest for dataloader with restrictions.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/vad_utils.py:100]: what do restrictions mean here?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 3]
```diff
@@ -27,75 +27,272 @@
...
+    """
+    specEmbA, specEmbB = specEmbA.to(device), specEmbB.to(device)
+    A, B = specEmbA.unsqueeze(dim=1), specEmbB.unsqueeze(dim=0)
+    dis = (A - B) ** 2.0
+    dis = dis.sum(dim=-1).squeeze()
+    return dis
+
+
+@torch.jit.script
+def kmeans_plusplus_torch(
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:98]: Good point. This part is mainly followed scikit learn and a few papers so I think we should mention it.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 4]
```diff
@@ -27,75 +27,272 @@
...
+    X, n_clusters: int, random_state: int, n_local_trials: int = 20, device: torch.device = torch.device('cpu')
+):
+    """
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:101]: Makes sense. I will add some details about kmeans plusplus algorithm

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:98]: All of the k-means parts are influenced by https://github.com/scikit-learn/scikit-learn. I will attach the paper info with scikit-learn.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:101]: makes sense. At least I could explain the motivation of k-means plusplus.

File: nemo/collections/asr/parts/utils/vad_utils.py [hunk 5]
```diff
@@ -224,166 +247,176 @@ def generate_overlap_vad_seq(
...
-            nan_idx = np.isnan(preds)
-            last_non_nan_pred = preds[~nan_idx][-1]
-            preds[nan_idx] = last_non_nan_pred
+    overlap = per_args['overlap']
+    window_length_in_sec = per_args['window_length_in_sec']
+    shift_length_in_sec = per_args['shift_length_in_sec']
 
-        else:
-            raise ValueError("smoothing_method should be either mean or median")
+    shift = int(shift_length_in_sec / 0.01)  # number of units of shift
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/parts/utils/vad_utils.py:279]: @tango4j  Good point. This is actually a bad practice. I will refactor this part thoroughy later.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 6]
```diff
@@ -27,75 +27,272 @@
...
+        if center_shift ** 2 < threshold:
+            break
+        if iter_limit > 0 and iter_count >= iter_limit:
+            break
+
+    return selected_cluster_index
+
+
+@torch.jit.script
+def getTheLargestComponent(affinity_mat, seg_index: int, device: torch.device):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:234]: torch.jit.script compiler assumes that everything is torch.Tensor, if not specified!!  Very arrogant compiler. So I just left the variables that are torch.Tensor. ->  I will specify the type hint for all variables and check if it works well.

File: nemo/collections/asr/parts/utils/vad_utils.py (hunk 2 above)
Comment (fayejf) [reply nemo/collections/asr/parts/utils/vad_utils.py:100]: oh good point. it means with at most split_duration duration. Let me clarify it.

File: nemo/collections/asr/parts/utils/vad_utils.py [hunk 7]
```diff
@@ -471,90 +535,154 @@ def filtering(speech_segments, per_args):
...
+            0, 10, speech
+            17,18, speech
+    Args:
+        vad_pred_dir (str): directory of prediction files to be processed.
+        postprocessing_params (dict): dictionary of thresholds for prediction score. See details in binarization and filtering.
+        shift_length_in_sec (float): amount of shift of window for generating the frame.
+        out_dir (str): output dir of generated table/csv file.
+        num_workers(float): number of process for multiprocessing
+    Returns:
+        table_out_dir(str): directory of generate table.
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/vad_utils.py]: There is a similar error still hasn't fixed. https://github.com/NVIDIA/NeMo/blob/d78755546105799e0e1423052656f9261592b4df/nemo/collections/asr/parts/utils/vad_utils.py#L236

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py [hunk 8]
```diff
@@ -670,7 +665,7 @@ def run_ASR_BPE_CTC(self, asr_model: Type[EncDecCTCModelBPE]) -> Tuple[Dict, Dic
                     hyp_words, word_ts = self.run_pyctcdecode(log_prob, onset_delay_in_sec=onset_delay_in_sec)
                 else:
                     logits_len = torch.from_numpy(np.array([len(greedy_predictions_list)]))
-                    greedy_predictions_list = greedy_predictions_list[onset_delay:-mid_delay]
+                    greedy_predictions_list = greedy_predictions_list[onset_delay:]
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:668]: can you elaborate why this change?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 9]
```diff
@@ -27,75 +27,312 @@
...
+    # Convert tensor type to float
+    X = X.float().to(device)
+    input_size = X.shape[0]
+
+    # Initialize the cluster centers with kmeans_plusplus algorithm.
+    plusplus_init_states = kmeans_plusplus_torch(X, n_clusters=num_clusters, random_state=0, device=device)
+    centers = plusplus_init_states[0]
+
+    iter_count = 0
+    selected_cluster_indices = torch.zeros(input_size).int()
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:236]: I don't get it. why converting above to float and use int here?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 10]
```diff
@@ -154,218 +389,282 @@ def get_argmin_mat(uniq_scale_dict):
...
-    def __init__(self, n_clusters=8, random_state=0, n_init=10, p_value=10, n_jobs=None, cuda=False):
+@torch.jit.script
+class SpectralClustering:
+    def __init__(
+        self,
+        n_clusters: int = 8,
+        random_state: int = 0,
+        n_init: int = 10,
+        p_value: int = 10,
+        cuda: bool = False,
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:633]: Seems like  cuda and is_cuda are the same thing. Can we unify them?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 11]
```diff
@@ -154,218 +389,282 @@ def get_argmin_mat(uniq_scale_dict):
...
         laplacian = getLaplacian(affinity_mat)
         lambdas_, diffusion_map_ = eigDecompose(laplacian, cuda)
         diffusion_map = diffusion_map_[:, :n_spks]
-        embedding = diffusion_map.T[n_spks::-1]
+        inv_idx = torch.arange(diffusion_map.size(1) - 1, -1, -1).long()
+        embedding = diffusion_map.T[inv_idx, :]
         return embedding[:n_spks].T
 
 
+@torch.jit.script
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:667]: oh I didn't know we can @torch.jit.script the whole class. Does it mean all functions within this class is torch-jit-scriptable?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 12]
```diff
@@ -619,7 +924,7 @@ def COSclustering(
             Default is 0.25.
```
Comment (fayejf) [inline nemo/collections/asr/parts/utils/nmesc_clustering.py:924]: you set the default to be 0.15 above. is it 0.25 or 0.15?

**Review by fayejf — COMMENTED** [review]
Great work. One general suggestion: since you wrote input type and your output is not ambiguous, how about adding type annotation for output as well?

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:98]: yes, please do

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:105]: Also, may be we can explain how this slightly differs from sk learn implementation

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py]: that would be perfect

**Comment by fayejf** [comment]
@tango4j  Forget to mention please make sure to add notes about the difference of the current implementation and the previous one.

File: nemo/collections/asr/parts/utils/decoder_timestamps_utils.py (hunk 8 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/decoder_timestamps_utils.py:668]: This was fixed because of (this issue)[https://github.com/NVIDIA/NeMo/issues/3878]. Sorry that I missed this update while commenting.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 9 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:236]: selected_cluster_indices is cluster index as the name suggests. So it needs to be int. While spectral embedding X is converted to float() to match the type with other operations.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py (hunk 12 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:919]: Good catch. Changing it to 0.15.

**Review by jbalam-nv — APPROVED** [review]
Minor comments but looks good overall.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 13]
```diff
@@ -377,51 +674,55 @@ class NMESC:
     convert the rest of values to 0.
 
     p_value can be also tuned on a development set without performing
-    NME-analysis.
+    NME-analysis. This will brings about significantly faster clustering speed,
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:731]: Fixed this to make it grammatically correct.

File: nemo/collections/asr/parts/utils/nmesc_clustering.py [hunk 14]
```diff
@@ -27,75 +27,310 @@
...
 
 from collections import Counter
+from typing import Dict, List
 
-import numpy as np
 import torch
-from sklearn.cluster._kmeans import k_means
-from sklearn.metrics.pairwise import cosine_similarity
-from sklearn.preprocessing import MinMaxScaler
+from torch.linalg import eigh as eigh
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmesc_clustering.py:38]: This is redundant. Removed.

**Review by nithinraok — DISMISSED** [review]

**Review by jbalam-nv — APPROVED** [review]
LGTM

**Review by nithinraok — APPROVED** [review]
